### ISSUE TITLE
Mismatch in valid samples mask

### ISSUE THREAD

**Issue by arbaazkhan2** [waymo-research/waymo-open-dataset#302]
While pre-processing the dataset with the 'sample_is_valid' flag we found a discrepancy that seems to be throwing off our model; 

The input_states in the demo is set to [128 x 11 x 2] The 11 comes from 10 samples from 1 sec of history + the current state. The 2 represents x and y positions. 
After masking out invalid agents; the input is [N x 11 x 2] (for example the first dataset (training_tfexample.tfrecord-00000-of-01000) in the training dir yields 65 x 11 x 2 when batch_size is set to 1). 

However, on closer inspection, we found that not all 65 agents were valid and in fact some of them only had valid x/y positions for some of the timestamps and the ground truth were set to [-1, -1] . For example, when inspecting state for agent index 64 in the above dataset, we found that 
the input_state was : 
tf.Tensor(
[[ 1.6429065e+03 -7.2779087e+03]
 [-1.0000000e+00 -1.0000000e+00]
 [-1.0000000e+00 -1.0000000e+00]
 [-1.0000000e+00 -1.0000000e+00]
 [-1.0000000e+00 -1.0000000e+00]
 [-1.0000000e+00 -1.0000000e+00]
 [-1.0000000e+00 -1.0000000e+00]
 [-1.0000000e+00 -1.0000000e+00]
 [-1.0000000e+00 -1.0000000e+00]
 [-1.0000000e+00 -1.0000000e+00]
 [-1.0000000e+00 -1.0000000e+00]], shape=(11, 2), dtype=float32)

and the ground truth was [-1, -1] for all of ground_truth[64,0:80,2]. This agent is clearly invalid but somehow the masking doesn't work on it (mostly because the masking is based on "If a sample was not seen at all in the past, we declare the sample as invalid."). How then do we handle agents which were seen in the past but disappear in the future? 

While we haven't inspected the 'tracks_to_predict' mask carefully is it possible that the tracks_to_predict mask also has similar agents that are visible only up to a certain point and then disappear in the future? If not would this not be a better mask than the samples_is_valid mask?

Is this by design or something unintentional? We could code in a filter to remove any agents that disappear in the future but obviously, if the prediction metric is looking at such an agent's predicted output, then this would throw off the results from the model. Any insight into this would be appreciated.

**Comment by scott-ettinger** [comment]
Hi Arbaaz, 

I'm not sure I fully understand how you are masking agents.  There are 3 different per-timestep valid masks that indicate when an agent has valid data.  The state/past/valid tensor indicates for each past time step which of the 128 agents has valid data.  Similarly, the state/current/valid tensor indicates which agents have valid data at the current time step and the state/future/valid tensor indicates which of the 128 agents have valid data in each future time step.  

The tracks_to_predict tensor provides a list of which agents your model must predict for the challenge for each scenario. These agents are guaranteed to be valid at the current time step and have valid states for a significant percentage of both the past and future time steps. 

The full documentation for the tensors mentioned above can be found here : https://waymo.com/open/data/motion/tfexample.

Please let me know if this does not answer your question.

**Comment by arbaazkhan2** [comment]
Hi Scott, the masking strategy I follow is the same one that's in the demo file;

past_is_valid = decoded_example['state/past/valid'] > 0
current_is_valid = decoded_example['state/current/valid'] > 0
sample_is_valid = tf.reduce_any( tf.concat([past_is_valid, current_is_valid], 1), 1)

and then 
states = tf.boolean_mask(inputs['input_states'], sample_is_valid)*.
 
This masking strategy as outlined in the demo is what is resulting in a bunch of input states where the input is only valid for a small part of the history time horizon. 

Thanks for confirming that the tracks_to_predict tensor provides agents with guaranteed valid states; 

*I am using the same variable names as in the demo colab.

**Comment by yuningchai** [comment]
Hi there, I am very sorry for the late reply.

Actually, we found an unrelated bug, and we are changing the colab tutorial in the following days, where we'll be using `tracks_to_predict` to identify valid tracks instead. That will probably fix this issue by chance.

**Comment by yuningchai** [comment]
It has been updated. Please verify if the masking works as expected with the new tracks_to_predict field. Thanks!

**Comment by arbaazkhan2** [comment]
Hey, 
I see the updated  colab; I have been using input states masked with tracks to predict and it seems to be working fine. However, 
when I generate the submission on the validation set, the server returns the error (I am only predicting outputs for agents per scenario identified by the tracks_to_predict mask)

CANCELLED: Invalid predicted trajectory length. Expected : 16 found : 0 trajectory : object_id: 1637

It's highly likely there is a bug in my code, but is it possible that there is a bug in this? (i.e the object_id is not in the tracks_to_predict mask for whatever scenario/tf.record file object 1637 lives in, but it is present in the online checker?)
